# Bitcoin historical data analysis

- Sơ lược về dataset( lấy ở đâu, và những điều mình sẽ làm để sử lý giữ liệu)
- tại sao Bitcoin là một tài sản lưu trữ giá trị
- các tool mà mình sử dụng như (pandas, matplotlib, seaborn)
- Những điều bản thân học được từ khóa học [Data Analysis with Python: Zero to Pandas](zerotopandas.com)


## Downloading the Dataset

**TODO** - add some explanation here

> Instructions for downloading the dataset (delete this cell)
>
> - Find an interesting dataset on this page: https://www.kaggle.com/datasets?fileType=csv
> - The data should be in CSV format, and should contain at least 3 columns and 150 rows
> - Download the dataset using the [`opendatasets` Python library](https://github.com/JovianML/opendatasets#opendatasets)

In [1]:
!pip install jovian opendatasets --upgrade --quiet

Let's begin by downloading the data, and listing the files within the dataset.

In [2]:
# copy and paste the link of dataset on Kaggle
dataset_url = 'https://www.kaggle.com/datasets/mczielinski/bitcoin-historical-data' 

In [3]:
import opendatasets as od
od.download(dataset_url)

Please provide your Kaggle credentials to download this dataset. Learn more: http://bit.ly/kaggle-creds
Your Kaggle username: huydinhcm
Your Kaggle Key: ········


100%|██████████| 100M/100M [00:01<00:00, 61.7MB/s] 


The dataset has been downloaded and extracted.

In [4]:
data_dir = './bitcoin-historical-data'

In [5]:
import os
os.listdir(data_dir)

['bitstampUSD_1-min_data_2012-01-01_to_2021-03-31.csv']

Let us save and upload our work to Jovian before continuing.

In [6]:
project_name = "bitcoin-historical-data-analysis" 

In [7]:
!pip install jovian --upgrade -q

In [8]:
import jovian

In [ ]:
jovian.commit(project=project_name)

<IPython.core.display.Javascript object>

## Data Preparation and Cleaning

**TODO** - Write some explanation here.



> Instructions (delete this cell):
>
> - Load the dataset into a data frame using Pandas
> - Explore the number of rows & columns, ranges of values etc.
> - Handle missing, incorrect and invalid data
> - Perform any additional steps (parsing dates, creating additional columns, merging multiple dataset etc.)

In [10]:
import pandas as pd

In [11]:
bitcoin_df = pd.read_csv(data_dir + "/bitstampUSD_1-min_data_2012-01-01_to_2021-03-31.csv")

In [12]:
bitcoin_df

Timestamp      Open      High       Low     Close  Volume_(BTC)  \
0        1325317920      4.39      4.39      4.39      4.39      0.455581   
1        1325317980       NaN       NaN       NaN       NaN           NaN   
2        1325318040       NaN       NaN       NaN       NaN           NaN   
3        1325318100       NaN       NaN       NaN       NaN           NaN   
4        1325318160       NaN       NaN       NaN       NaN           NaN   
...             ...       ...       ...       ...       ...           ...   
4857372  1617148560  58714.31  58714.31  58686.00  58686.00      1.384487   
4857373  1617148620  58683.97  58693.43  58683.97  58685.81      7.294848   
4857374  1617148680  58693.43  58723.84  58693.43  58723.84      1.705682   
4857375  1617148740  58742.18  58770.38  58742.18  58760.59      0.720415   
4857376  1617148800  58767.75  58778.18  58755.97  58778.18      2.712831   

         Volume_(Currency)  Weighted_Price  
0                 2.000000        4.390000  
1                      NaN             NaN  
2                      NaN             NaN  
3                      NaN             NaN  
4                      NaN             NaN  
...                    ...             ...  
4857372       81259.372187    58692.753339  
4857373      428158.146640    58693.226508  
4857374      100117.070370    58696.198496  
4857375       42332.958633    58761.866202  
4857376      159417.751000    58764.349363  

[4857377 rows x 8 columns]

Có 8 cột và 4,857,377 hàng theo thông tin của dataframe info. giờ tôi sẽ sem thử 10 hàng đầu tiên của data frame và sẽ bắt đầu các bước sử lý với Timestamp và các NaN values.

In [13]:
bitcoin_df['Date'] = pd.to_datetime(bitcoin_df['Timestamp'], unit='s')

In [14]:
bitcoin_df['Date']

0         2011-12-31 07:52:00
1         2011-12-31 07:53:00
2         2011-12-31 07:54:00
3         2011-12-31 07:55:00
4         2011-12-31 07:56:00
                  ...        
4857372   2021-03-30 23:56:00
4857373   2021-03-30 23:57:00
4857374   2021-03-30 23:58:00
4857375   2021-03-30 23:59:00
4857376   2021-03-31 00:00:00
Name: Date, Length: 4857377, dtype: datetime64[ns]

In [15]:
bitcoin_df.loc[0, 'Date'].day_name()

'Saturday'

In [16]:
bitcoin_df['Date'].dt.day_name()

0           Saturday
1           Saturday
2           Saturday
3           Saturday
4           Saturday
             ...    
4857372      Tuesday
4857373      Tuesday
4857374      Tuesday
4857375      Tuesday
4857376    Wednesday
Name: Date, Length: 4857377, dtype: object

In [17]:
bitcoin_df['Day_Of_Week']= bitcoin_df['Date'].dt.day_name()

In [18]:
bitcoin_df.pop("Timestamp")

0          1325317920
1          1325317980
2          1325318040
3          1325318100
4          1325318160
              ...    
4857372    1617148560
4857373    1617148620
4857374    1617148680
4857375    1617148740
4857376    1617148800
Name: Timestamp, Length: 4857377, dtype: int64

In [19]:
bitcoin_df.head(5)

Open  High   Low  Close  Volume_(BTC)  Volume_(Currency)  Weighted_Price  \
0  4.39  4.39  4.39   4.39      0.455581                2.0            4.39   
1   NaN   NaN   NaN    NaN           NaN                NaN             NaN   
2   NaN   NaN   NaN    NaN           NaN                NaN             NaN   
3   NaN   NaN   NaN    NaN           NaN                NaN             NaN   
4   NaN   NaN   NaN    NaN           NaN                NaN             NaN   

                 Date Day_Of_Week  
0 2011-12-31 07:52:00    Saturday  
1 2011-12-31 07:53:00    Saturday  
2 2011-12-31 07:54:00    Saturday  
3 2011-12-31 07:55:00    Saturday  
4 2011-12-31 07:56:00    Saturday

In [20]:
bitcoin_df['Date'].min()

Timestamp('2011-12-31 07:52:00')

In [21]:
bitcoin_df['Date'].max()

Timestamp('2021-03-31 00:00:00')

In [22]:
bitcoin_df['Date'].max() - bitcoin_df['Date'].min()

Timedelta('3377 days 16:08:00')

Có khá nhiều 'Not a Number' values, và đầu tiên tôi sẽ sử lý timestamp để dễ hiểu hơn và hỗ trợ dư liệu dễ hiểu hơn.

In [22]:
import jovian

In [23]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "aakashns/zerotopandas-course-project-starter" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/aakashns/zerotopandas-course-project-starter


'https://jovian.ml/aakashns/zerotopandas-course-project-starter'

## Exploratory Analysis and Visualization

**TODO** - write some explanation here.



> Instructions (delete this cell)
> 
> - Compute the mean, sum, range and other interesting statistics for numeric columns
> - Explore distributions of numeric columns using histograms etc.
> - Explore relationship between columns using scatter plots, bar charts etc.
> - Make a note of interesting insights from the exploratory analysis

Let's begin by importing`matplotlib.pyplot` and `seaborn`.

In [24]:
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline

sns.set_style('darkgrid')
matplotlib.rcParams['font.size'] = 14
matplotlib.rcParams['figure.figsize'] = (9, 5)
matplotlib.rcParams['figure.facecolor'] = '#00000000'

**TODO** - Explore one or more columns by plotting a graph below, and add some explanation about it

**TODO** - Explore one or more columns by plotting a graph below, and add some explanation about it

**TODO** - Explore one or more columns by plotting a graph below, and add some explanation about it

**TODO** - Explore one or more columns by plotting a graph below, and add some explanation about it

**TODO** - Explore one or more columns by plotting a graph below, and add some explanation about it

Let us save and upload our work to Jovian before continuing

In [25]:
import jovian

In [26]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "aakashns/zerotopandas-course-project-starter" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/aakashns/zerotopandas-course-project-starter


'https://jovian.ml/aakashns/zerotopandas-course-project-starter'

## Asking and Answering Questions

TODO - write some explanation here.



> Instructions (delete this cell)
>
> - Ask at least 5 interesting questions about your dataset
> - Answer the questions either by computing the results using Numpy/Pandas or by plotting graphs using Matplotlib/Seaborn
> - Create new columns, merge multiple dataset and perform grouping/aggregation wherever necessary
> - Wherever you're using a library function from Pandas/Numpy/Matplotlib etc. explain briefly what it does



#### Q1: TODO - ask a question here and answer it below

#### Q2: TODO - ask a question here and answer it below

#### Q3: TODO - ask a question here and answer it below

#### Q4: TODO - ask a question here and answer it below

#### Q5: TODO - ask a question here and answer it below

Let us save and upload our work to Jovian before continuing.

In [28]:
import jovian

In [29]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "aakashns/zerotopandas-course-project-starter" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/aakashns/zerotopandas-course-project-starter


'https://jovian.ml/aakashns/zerotopandas-course-project-starter'

## Inferences and Conclusion

**TODO** - Write some explanation here: a summary of all the inferences drawn from the analysis, and any conclusions you may have drawn by answering various questions.

In [30]:
import jovian

In [31]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "aakashns/zerotopandas-course-project-starter" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/aakashns/zerotopandas-course-project-starter


'https://jovian.ml/aakashns/zerotopandas-course-project-starter'

## References and Future Work

**TODO** - Write some explanation here: ideas for future projects using this dataset, and links to resources you found useful.

> Submission Instructions (delete this cell)
> 
> - Upload your notebook to your Jovian.ml profile using `jovian.commit`.
> - **Make a submission here**: https://jovian.ml/learn/data-analysis-with-python-zero-to-pandas/assignment/course-project
> - Share your work on the forum: https://jovian.ml/forum/t/course-project-on-exploratory-data-analysis-discuss-and-share-your-work/11684
> - Share your work on social media (Twitter, LinkedIn, Telegram etc.) and tag [@JovianML](https://twitter.com/jovianml)
>
> (Optional) Write a blog post
> 
> - A blog post is a great way to present and showcase your work.  
> - Sign up on [Medium.com](https://medium.com) to write a blog post for your project.
> - Copy over the explanations from your Jupyter notebook into your blog post, and [embed code cells & outputs](https://medium.com/jovianml/share-and-embed-jupyter-notebooks-online-with-jovian-ml-df709a03064e)
> - Check out the Jovian.ml Medium publication for inspiration: https://medium.com/jovianml


 

In [32]:
import jovian

In [35]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "aakashns/zerotopandas-course-project-starter" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/aakashns/zerotopandas-course-project-starter


'https://jovian.ml/aakashns/zerotopandas-course-project-starter'